cpp/dolfinx/fem/utils.h

```diff
@@ -757,82 +698,118 @@ Form<T, U> create_form_factory(
     }
   }
 
-  // Attach vertex kernels
+  // Attach exterior entity integrals
   {
-    for (std::size_t form_idx = 0; form_idx < ufcx_forms.size(); ++form_idx)
+    for (IntegralType itg_type : {IntegralType::exterior_facet,
+                                  IntegralType::vertex, IntegralType::ridge})
     {
-      const ufcx_form& form = ufcx_forms[form_idx];
-
-      std::span<const int> ids(form.form_integral_ids
-                                   + integral_offsets[vertex],
-                               num_integrals_type[vertex]);
-      auto sd = subdomains.find(IntegralType::vertex);
-      for (int i = 0; i < num_integrals_type[vertex]; ++i)
+      std::size_t dim;
+      switch (itg_type)
       {
-        const int id = ids[i];
-        ufcx_integral* integral
-            = form.form_integrals[integral_offsets[vertex] + i];
-        assert(integral);
-        check_geometry_hash(*integral, form_idx);
+      case IntegralType::exterior_facet:
+      {
+        dim = tdim - 1;
+        break;
+      }
+      case IntegralType::ridge:
+      {
+        dim = tdim - 2;
+        break;
+      }
+      case IntegralType::vertex:
+      {
+        dim = 0;
+        break;
+      }
+      default:
+        throw std::runtime_error("Unsupported integral type");
+      }
 
-        std::vector<int> active_coeffs;
-        for (int j = 0; j < form.num_coefficients; ++j)
+      auto get_default_integration_entities
+          = [dim](const mesh::Topology& topology, IntegralType itg_type)
+      {
+        if (itg_type == IntegralType::exterior_facet)
         {
-          if (integral->enabled_coefficients[j])
-            active_coeffs.push_back(j);
+          // Integrate over all owned exterior facets
+          return mesh::exterior_facet_indices(topology);
+        }
+        else
+        {
+          // Integrate over all owned entities
+          // Default vertex kernel operates on all (owned) vertices
```

Comment:
What does this mean here? Can ridges be captured in this block?

Reply:
Ah, this is a comment that remains from the vertex integration kernel. This should just be the first comment.
It captures cells, facets, ridges and vertices.